### ISSUE TITLE
Filtering not working for multi value URL parameters (GET request)

### ISSUE THREAD

**Issue by reygoch** [symphonycms/symphony-2#2616]
Let's say I have a section with tags, and another section with multiple select box that selects from tags.

If I want to filter that section by multiple tags, I'd do something like this :

    ?tags[]=blue&tags[]=green&tags[]=red

This gives me following nodes :

    <url-tags.1>blue</url-tags.1>
    <url-tags.2>green</url-tags.2>
    <url-tags.3>red</url-tags.3>

And following parameters :

    $url-tags.1    'blue'
    $url-tags.2    'green'
    $url-tags.3    'red'

I can't really filter DS with that, since I never know how many tags I will have, so I can't do something like `{$url-tags.1:}, {$url-tags.2:}, {$url-tags.3:}` as filter criteria. This also doesn't work `{$url-tags}` (it was my last hope) :D

This also doesn't work it gives similar results to the ones above :

    ?fields[tags][]=Blue&fields[tags][]=Brown

I suggest that generated xml for parameters look like this :

    <url-tags>
        <value>tag01</value>
        <value>tag02</value>
    </url-tags>

Or for more complex url parameter :

    ?field-group[field1][]=f1v1&field-group[field1][]=f1v2&field-group[field1][]=f2v1&field-group[field2][]=f2v2

XML :

    <url-field-group>
        <url-field1>
            <val>f1v1</val>
            <val>f1v2</val>
        </url-field1>
        <url-field2>
            <val>f2v1</val>
            <val>f2v2</val>
        </url-field2>
    </url-field-group>

**Comment by johnpuddephatt** [comment]
I think this is closely related to #1807 and  #2484 ?

**Comment by reygoch** [comment]
Also, following parameters should be created :

```
$url-tags = blue, green, red
$url-tags-and = blue+green+red

$url-field-group.url-field1 = f1v1, f1v2
$url-field-group.url-field1-and = f1v1+f1v2

$url-field-group.url-field2 = f2v1, f2v2
$url-field-group.url-field2-and = f2v1+f2v2

```

**Comment by reygoch** [comment]
@johnpuddephatt yes, but the seem kind of forgotten ;D

**Comment by nitriques** [comment]
> Also, following parameters should be created :

Where does the `-and` part comes from ?

You shoud use `tag=1,2,3` instead of php's array syntax...

**Comment by reygoch** [comment]
@nitriques comma separated values are treated as `OR` while plus separated values are treated as `AND`. It would be practical to have both of them to allow us for better filtering.

I'm not sure what you mean by php array syntax. This maybe `tags[]=value`? This is part of URL that form generates once you have submitted value from multi select box.

Forms don't generate URLs with comma separated values. This can be achieved with JavaScript but should be possible with vanilla Symphony.

**Comment by nitriques** [comment]
>  It would be practical to have both of them to allow us for better filtering.

You only need one at the time, I do not see why both are needed. You can url-encode the `+` sign to get the AND.

> This maybe tags[]=value?

Yup

> This is part of URL that form generates once you have submitted value from multi select box.

Only if the field name contains `[]`, which you can remove.

> Forms don't generate URLs with comma separated values.

Symphony generated forms are meant to save data into sections, not do filtering. I guess that you need to do some javascript to make it work.

I'd like to get @brendo's feedback on this, as I am not well placed to discuss this part of the system ;)

**Comment by reygoch** [comment]
no, no, no... I think we have a misunderstanding here :D @nitriques

I'm talking about getting adhoc parameters, not section events or page parameters. And `[]` is kind of a standard notation for fields with multiple values so I just put them there so that people can understand what's going on there.

**Comment by nitriques** [comment]
> I think we have a misunderstanding here

Maybe, please explain ;)

> And [] is kind of a standard notation for fields with multiple values 

I agree. But right now, data sources are not meant to work with this...

**Comment by michael-e** [comment]
I remember a discussion (long time ago) about this, and a simple solution: Add an event to your page to manipulate the form data. Here is an example event:

```php
<?php

require_once(TOOLKIT.'/class.event.php');

class eventcomma_separate_get_params extends Event
{
    public $ROOTELEMENT = 'comma-separate-get-params';

    public $eParamFILTERS = array(
    );

    public static function about()
    {
        return array(
                'name' =>'Comma separate $_GET params',
                'author' => array(
                    'name' => 'John Porter, Michael Eichelsdoerfer'
                ),
                'version' => '1.0',
                'release-date' => '2012-10-16'
        );
    }

    public static function allowEditorToParse()
    {
        return false;
    }

    public static function documentation()
    {
        return '';
    }

    public function load()
    {
        return $this->__trigger();
    }

    protected function __trigger()
    {
        $page = Frontend::instance()->Page();
        if (is_array($_GET) && !empty($_GET)) {
            foreach ($_GET as $key => $value) {
                if (is_array($value)) {
                    $page->_param['url-'.$key] = join(',', $value);
                }
            }
        }
    }
}
```

**Comment by johnpuddephatt** [comment]
> I remember a discussion (long time ago) about this

[only three years](https://github.com/symphonycms/symphony-2/issues/1807#issuecomment-23156310) ;)

It does look like some changes were agreed in #1807 and #2484 but never happened.

**Comment by reygoch** [comment]
@nitriques well... yes, currently data sources don't work with that and this is exactly what I'm talking about :D

**Comment by reygoch** [comment]
@nitriques no, because html forms don't generate comma separated values on multiselect fields. I can do it with JavaScript but it should be possible with vanilla symphony setup.

**Comment by nitriques** [comment]
> but it should be possible with vanilla symphony setup.

Right now, it's not. Like I've said, if you want to make it work right now, you must do js.

We'll work on making it possible tho, but that could take a long time.

**Comment by reygoch** [comment]
@nitriques as I said, I know that :D

**Comment by nitriques** [comment]
Ok perfect ;) I am not an native English speaker so I tend to repeat myself ;)

**Comment by jensscherbl** [comment]
> I suggest that generated xml for parameters look like this :

    <url-tags>
        <value>tag01</value>
        <value>tag02</value>
    </url-tags>

If you still need it, I have an unreleased extension for this. Not sure if it works with your more complex example, though.

**Comment by reygoch** [comment]
@jensscherbl thanks, although I'm kind of in the middle of making a little promo / tutorial for symphony and this doesn't look very good :D

**Comment by jensscherbl** [comment]
> although I'm kind of in the middle of making a little promo / tutorial for symphony and this doesn't look very good :D

Agreed. It's already on the list for the `3.0.0` release.
